content/post/2021-07-29-caching/index.Rmd

```diff
@@ -0,0 +1,294 @@
+```
+
+In the whoami package by Gábor Csárdi, there is an internal function `lookup_gh_username()` that calls the GitHub API. It is memoized without the memoise package.
+
+-   In [`.onLoad()`](https://github.com/r-lib/whoami/blob/40999c9945104f740d0fe13ed07288879aec14c6/R/whoami.R#L2) the memoization function is called
+
+``` r
+.onLoad <- function(libname, pkgname) {
+  lookup_gh_username <<- memoize_first(lookup_gh_username)
+}
+```
+
+-   What's the [memoization function](https://github.com/r-lib/whoami/blob/40999c9945104f740d0fe13ed07288879aec14c6/R/whoami.R#L10), you ask?
+
+``` r
+memoize_first <- function(fun) {
+  fun
+  cache <- list()
+  dec <- function(arg, ...) {
+    if (!is_string(arg)) return(fun(arg, ...))
+    if (is.null(cache[[arg]])) cache[[arg]] <<- fun(arg, ...)
+    cache[[arg]]
+  }
+  dec
+}
+```
+
+This function is a closure (a [function creating a function](https://adv-r.hadley.nz/function-factories.html?q=closure#function-factories)).
+It caches results in a list from where they are retrieved in the function gets called a second time with the same argument.
+
+In the [Advanced R book](https://adv-r.hadley.nz/function-factories.html?q=closure#stateful-funs) by Hadley Wickham such function factories are called *stateful functions* that "allow you to maintain state across function invocations".
+It also has a warning on not abusing them.
+
+> "Stateful functions are best used in moderation. As soon as your function starts managing the state of multiple variables, it's better to switch to R6, the topic of Chapter 14."
+
+### Saving results in an environment
+
+This is well suited for package development where:
+
+-   You create an environment internal to your package where you store a value;
+-   You can then store in it any computed value for the current R session where the package is loaded.
+
+Here's an example with only base R functions:
+
+``` r
+cache_env <- new.env(parent = emptyenv())
+who_am_i <- function() {
+  if (is.null(cache_env$name)) {
+    cache_env$name <- readline("I don't know you. Can you tell me your name ? ")
+  }
+  message("Welcome ", cache_env$name, "!")
+  invisible(cache_env$name)
+}
+get_current_name <- function() {
+  cache_env$name
+}
+who_am_i()
+#> I don't know you. Can you tell me your name ? cderv
+#> Welcome cderv!
+who_am_i()
+#> Welcome cderv!
+```
+
+This example would be simpler if using `rlang::env_cache()` by Lionel Henry (in the development version of [rlang](https://rlang.r-lib.org/) at the time of writing).
+
+``` r
+cache_env <- rlang::new_environment()
+who_am_i <- function() {
+  name <- rlang::env_cache(
+    env = cache_env, nm = "name", 
+    default = readline("I don't know you. Can you tell me your name ? ")
+  )
+  message("Welcome ", name, "!")
+  invisible(name)
+}
+get_current_name <- function() {
+  cache_env$name
+}
+who_am_i()
+#> I don't know you. Can you tell me your name ? cderv
+#> Welcome cderv!
+who_am_i()
+#> Welcome cderv!
+```
+
+### Storing on disk?
+
```

Comment:
Should we mention some cache  specific package for those who wants to go further in this topic ? 

I have in mind 

* storr https://cran.r-project.org/web/packages/storr/index.html
* R.cache https://cran.r-project.org/web/packages/R.cache/
* cachem https://cran.r-project.org/web/packages/cachem